From 42c0ca338a705a8122be79997018676d018c7b09 Mon Sep 17 00:00:00 2001
From: balajhidn <balajhix.neelakantan.durvas@intel.com>
Date: Thu, 15 Sep 2022 14:52:50 +0000
Subject: [PATCH] Secure boot flash partition update

For ROT image type (0xf1),
    ROT/ SPL block update to primary partition (Image-A).
    U-boot, PFM and Kernel Fit image update to passive partition.

For BMC image type (0x04),
    ROT/ SPL block, U-boot, PFM and Kernel Fit image update to passive partition.

Tested:
    Firmware update for both partitions are success.
    BMC booted successfully in both cases.

Signed-off-by: Gade-Rajasekhar Reddy <raja.sekhar.reddy.gade@intel.com>
Signed-off-by: balajhidn <balajhix.neelakantan.durvas@intel.com>
Upstream-Status: Pending
---
 board/aspeed/ast2600_intel/fw-update.c | 29 +++++++++++++++++++++++++-
 board/aspeed/ast2600_intel/fw-update.h |  1 +
 2 files changed, 29 insertions(+), 1 deletion(-)

diff --git a/board/aspeed/ast2600_intel/fw-update.c b/board/aspeed/ast2600_intel/fw-update.c
index dc2552f08c..85f9ebd291 100644
--- a/board/aspeed/ast2600_intel/fw-update.c
+++ b/board/aspeed/ast2600_intel/fw-update.c
@@ -809,7 +809,7 @@ int verify_spi_pfm(void)
 	ulong *buf = IMAGE_LOAD_RAM_ADDR;
 	ulong pfm_offset;
 
-	// Get active partition PFM start address
+	// Set PFM start address offset from active partition
 	if (get_boot_address() == SECONDARY_FIT_IMAGE_START_ADDR)
 	{
 		pfm_offset = SECONDARY_PFM_START_ADDR;
@@ -1137,7 +1137,11 @@ static int flash_image(void) {
 
 	u32 dev_offset = PRIMARY_IMAGE_OFFSET;
 	u32 blocks_to_skip = 0;
+	u8 pc_type;
+	u32 uboot_offset;
+	bool rot_a_update_done = false;
 
+	// When booted with active partition, FW update done on passive partition.
 	if (get_boot_address() == SECONDARY_FIT_IMAGE_START_ADDR)
 	{
 		dev_offset = PRIMARY_IMAGE_OFFSET;
@@ -1147,6 +1151,9 @@ static int flash_image(void) {
 		dev_offset = SECONDARY_IMAGE_OFFSET;
 	}
 
+	uboot_offset = dev_offset;
+	pc_type = get_pc_type();
+
 	// walk the bitmap, erase and copy
 	offset += (blk0blk1_size * 2); // one blk0blk1 for package, one for pfm signature
 
@@ -1182,6 +1189,26 @@ static int flash_image(void) {
 			blocks_to_skip = BLOCKS_SKIP;
 		}
 
+		// When secondary update is initiated for image type 0xf1 the ROT SPL is
+		// updated in the primary region and uboot, pfm, fit image is updated in
+		// secondary region
+		if (!rot_a_update_done)
+		{
+			if ((pc_type == pc_type_rot_update) &&
+				(uboot_offset == SECONDARY_IMAGE_OFFSET))
+			{
+				if ((blk == 0))
+				{
+					dev_offset = PRIMARY_IMAGE_OFFSET;
+				}
+				else if (blk == UBOOT_START_BLOCK)
+				{
+					dev_offset = uboot_offset;
+					rot_a_update_done = true;
+				}
+			}
+		}
+
 		if ((blk % 8) == 0)
 		{
 			wr_count = 1;
diff --git a/board/aspeed/ast2600_intel/fw-update.h b/board/aspeed/ast2600_intel/fw-update.h
index 591c79b0ae..f69150c8d7 100644
--- a/board/aspeed/ast2600_intel/fw-update.h
+++ b/board/aspeed/ast2600_intel/fw-update.h
@@ -19,6 +19,7 @@
 #define SECONDARY_IMAGE_OFFSET          0x04000000
 #define BLOCKS_SKIP                     0xA80
 #define FIT_IMAGE_BLOCK                 0xb00
+#define UBOOT_START_BLOCK               0x10
 
 #define MAX_FILENAME_LENGTH             256
 #define RAND_NUMBER_SIZE                8
-- 
2.25.1

